datafusion/core/src/physical_plan/topk/mod.rs

```diff
@@ -0,0 +1,516 @@
+                }
+            }
+        }
+        self.heap.insert_batch_entry(batch_entry);
+
+        // update memory reservation
+        self.reservation.try_resize(self.size())?;
+        Ok(())
+    }
+
+    /// Returns the top k results broken into `batch_size` [`RecordBatch`]es
+    pub fn emit(self) -> Result<SendableRecordBatchStream> {
+        let Self {
+            schema,
+            metrics,
+            reservation: _,
+            batch_size,
+            expr: _,
+            row_converter: _,
+            scratch_rows: _,
+            heap,
+        } = self;
+        let _timer = metrics.baseline.elapsed_compute().timer(); // time updated on drop
+
+        let mut batch = heap.emit(schema.clone())?;
+        metrics.baseline.output_rows().add(batch.num_rows());
+
+        // break into record batches as needed
+        let mut batches = vec![];
+        loop {
+            if batch.num_rows() < batch_size {
+                batches.push(Ok(batch));
+                break;
+            } else {
+                batches.push(Ok(batch.slice(0, batch_size)));
+                batch = batch.slice(batch_size, batch.num_rows());
+            }
+        }
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            schema,
+            futures::stream::iter(batches),
+        )))
+    }
+
+    /// return the size of memory used by this operator, in bytes
+    fn size(&self) -> usize {
+        std::mem::size_of::<Self>()
+            + self.row_converter.size()
+            + self.scratch_rows.size()
+            + self.heap.size()
+    }
+}
+
+struct TopKMetrics {
+    /// metrics
+    pub baseline: BaselineMetrics,
+
+    /// count of how many rows were replaced in the heap
+    pub row_replacements: Count,
+}
+
+impl TopKMetrics {
+    fn new(metrics: &ExecutionPlanMetricsSet, partition: usize) -> Self {
+        Self {
+            baseline: BaselineMetrics::new(metrics, partition),
+            row_replacements: MetricBuilder::new(metrics)
+                .counter("row_replacements", partition),
+        }
+    }
+}
+
+/// This structure keeps at most the *smallest* k items, using the
+/// [arrow::row] format for sort keys. While it is called "topK" for
+/// values like `1, 2, 3, 4, 5` the "top 3" really means the
+/// *smallest* 3 , `1, 2, 3`, not the *largest* 3 `3, 4, 5`.
+///
+/// Using the `Row` format handles things such as ascending vs
+/// descending and nulls first vs nulls last.
+///
+/// It doesn't use `BinaryHeap` in the Rust standard library because
+/// it is important to check the current minimum value in the heap
+/// prior to creating a new value to insert.
+struct TopKHeap {
+    /// The maximum size of this heap.
+    k: usize,
+    /// Storage for up at most `k` items, in ascending
+    /// order. `inner[0]` holds the smallest value of the smallest k
+    /// so far, `inner[len-1]` holds the largest value smallest k so far.
+    inner: Vec<TopKRow>,
```

Comment:
I feel like there must be some more clever algorithmic trick to keep a sorted list / balanced tree but this one seems to work pretty well 🤷 I do think it will have `O(N * K)` worst case performance (if the input is reverse ordered so every new values is  a new topK).

Reply:
https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html perhaps?

Reply:
The reason I couldn't figure out how to use a `BinaryHeap` is that I also need to get access to the *largest* (or smallest) value in the heap 

The way `BinaryHeap` seems to have been setup is that you push an owned something in it and then pull the top item. 

However, for this implementation it is critical not to have to actually construct an owned something unless it is a new top K (because constructing the item is expensive, given it contains an `OwnedRow`)

Reply:
BTW @avantgardnerio  suggested elsewhere that using a [BTreeSet](https://doc.rust-lang.org/stable/std/collections/struct.BTreeSet.htm) or similar structure would be better algorithmically (which we would see  larger `k` ) -- I'll give it a try over the next day or two

Reply:
> critical not to have to actually construct an owned something

So if I understand correctly, you are keeping OwnedRows in your `Vec`/`BTreeSet`, but you are doing a comparison like:

```
let new_row = Row;
let vec = SortedVec<OwnedRow>;
if new_row > vec.last().as_row() {
   // fail fast
}
```

If so, I think you could use the BinaryHeap<OwnedRow> the same way, just by calling `.peek()` to get the "worst" value.

> largest (or smallest) value in the heap

They have an [example](https://doc.rust-lang.org/stable/std/collections/struct.BinaryHeap.html#min-heap) of making it a min-heap with `Reverse`.

Reply:
I agree with @tustvold unless there's something funny going on with branch mispredicts or cache locality, the BinaryHeap should be the most efficient for this use-case.

I was unable to use the out-of-the box one because I am doing aggregation as well, which requires either pointers or random-lookups into the tree.

Reply:
🤔 I was writing a response but then I think you are right we could use BinaryHeap with `peek()`  -- I'll try that and see how it compares to a `BTreeSet` -- it might actually do very 🤔